core/src/main/java/org/apache/iceberg/util/Tasks.java

```diff
@@ -570,7 +584,7 @@ public static <I> Builder<I> foreach(I... items) {
   }
 
   public static <I> Builder<I> foreach(Stream<I> items) {
-    return new Builder<>(items::iterator);
+    return new Builder<>(items.iterator());
```

Comment:
I think a better fix is to introduce a one-time use `Iterable` that enforces `hasStarted` internally. That way we don't need to modify the code in `Tasks`, which is a much larger risk.

Reply:
> I think a better fix is to introduce a one-time use `Iterable` that enforces `hasStarted` internally.

i would assume `stream::iterator` does that. Still, it's not a real `Iterable` IMO.

Reply:
No, `stream::iterator` will allow calling `iterator` more than once. Adding a single-use iterable would fix this without lots of code changes.

Reply:
> No, `stream::iterator` will allow calling `iterator` more than once.

I didn't know that, thanks!

> Adding a single-use iterable would fix this 

That's true. But `Iterable` interface contract isn't clear about whether "a single-use iterable" is a valid implementation, so I wouldn't want to create such a class. I wouldn't stop you if you wanted to, though. 

> without lots of code changes. 

from long-term design perspective this may or may not be a dominant factor